doc/tutorials/content/sources/gpu_octree/gpu_octree.cpp

```diff
@@ -7,176 +7,160 @@
-        }
-    }
-    std::cout <<std::endl;
+  return std::make_pair(indices_h, sqr_dists_h);
+}
 
+auto
+radiusSearch(const pcl::gpu::Octree& octree,
+             const pcl::gpu::Octree::Queries& queries,
+             const std::vector<float>& radiuses_h)
+{
+  // prepare radiuses
+  pcl::gpu::Octree::Radiuses radiuses_d;
+  radiuses_d.upload(radiuses_h);
+
+  // prepare output buffers on device
+  const auto max_results = octree.cloud_->size();
+  pcl::gpu::NeighborIndices indices_d(queries.size(), max_results);
+
+  // search GPU
+  // uncomment once radius search PR is merged
+  // pcl::gpu::Octree::ResultSqrDists sqr_dists_d;
+  octree.radiusSearch(queries, radiuses_d, max_results, indices_d /*, sqr_dists_d*/);
+
+  // download results
+  std::vector<int> indices_h;
+  std::vector<int> sizes;
+  indices_d.data.download(indices_h);
+  // sqr_dists_d.download(sqr_dists_h);
+  indices_d.sizes.download(sizes);
+
+  // rewrite once radius search pr is merged
+  const std::vector<float> sqr_dists_h(indices_h.size(), NAN);
+  return std::make_tuple(indices_h, sizes, sqr_dists_h);
+}
 
-    //perform approximate nearest search
-    //prepare output buffers on device
-    pcl::gpu::NeighborIndices results_approx(query_size, 1);
-    pcl::gpu::Octree::ResultSqrDists dists_approx;
+auto
+approxNearestSearch(const pcl::gpu::Octree& octree,
+                    const pcl::gpu::Octree::Queries& queries)
+{
+  // prepare output buffers on device
+  pcl::gpu::NeighborIndices indices_d(queries.size(), 1);
+  pcl::gpu::Octree::ResultSqrDists sqr_dists_d;
 
-    //search GPU
-    octree_device.approxNearestSearch(queries_device, results_approx, dists_approx);
+  // search GPU
+  octree.approxNearestSearch(queries, indices_d, sqr_dists_d);
 
-    //download results
-    std::vector<int> downloaded_approx;
-    std::vector<float> dists_downloaded_approx;
-    results_approx.data.download(downloaded_approx);
-    dists_approx.download(dists_downloaded_approx);
+  // download results
+  std::vector<int> indices_h;
+  std::vector<float> sqr_dists_h;
+  indices_d.data.download(indices_h);
+  sqr_dists_d.download(sqr_dists_h);
 
-    for(std::size_t i = 0; i < query_size; ++i)
-    {
-        std::cout << "Approximate nearest neighbor search at (" << queries[i].x
-            << " " << queries[i].y
-            << " " << queries[i].z << ")" << std::endl;
+  return std::make_pair(indices_h, sqr_dists_h);
+}
 
-        std::cout << "\t" << cloud->points[downloaded_approx[i]].x
-            << " " << cloud->points[downloaded_approx[i]].y
-            << " " << cloud->points[downloaded_approx[i]].z
-            << " (squared distance " << dists_downloaded_approx[i] << ")" << std::endl;
+int
+main()
+{
+  using Generator =
+      pcl::common::CloudGenerator<pcl::PointXYZ, pcl::common::UniformGenerator<float>>;
+  Generator generator(Generator::GeneratorParameters(0, 1024.f)); // min max range
```

Comment:
We might need to limit this point generation to span a a plane. Notice what's happening with approx NN.
![appro](https://raw.githubusercontent.com/PointCloudLibrary/pcl/c72c1db0fe29307c21c0dc510014cd29a238dc8c/doc/tutorials/content/images/gpu/octree/gpu_approx.png)

the green dot (optimal)  looks further than the red dot (approximate, should be worse)  because of perspective. Making a pointcloud which spans a plane parallel to the image plane would get rid of this effect.

Reply:
Yeah I was wondering how to fix that. That should do the trick